report/41_Lessons-Evolution-&-Refactoring.md

```diff
@@ -1 +1,97 @@
+
+In the pipeline, the team has focused on having many actions done automatically, as mentioned in [TODO: 2.3: CI/CD](23_Process-CI-CD-Description.md).
+This was, at first, because the course demanded it, but once the team started using automatic steps, then the value was apparent.
+The pipeline was expanded repeatedly, with the steps themselves even being expanded, as is the case with `e2e.yml`.
+
+The takeaway here is that automation is very useful, as these manual, menial tasks could now be carried out automatically.
+A focus for the team members moving forward is to set up robust automation steps in future projects as soon as possible.
+
+#### Choose scalable technologies early
+
+The team knew that choosing scalable techonologies early was important.
+If a choice to switch from a monolithic server to a distributed system comes too late, then the changes that will have to be made could become quite big, resulting in a lot of technical debt.
+
+This was why dockerization came early for the project, with separation between components for easier clusterization.
+The team didn't quite get Docker Swarm to work, but because these considerations were done early, Docker Swarm was actually obtainable, instead of being buried in debt from not having a RESTful API, for example.
+
+#### Graphs are motivational
+
+After the simulator had been live for a few weeks, the statistics showed that this project lacked behind the other participating projects by quite a big margin.
+Having this as a motivator to improve was vital to finding some of the bugs in the system.
+
+First, it was discovered that the project lacked behind because response times were highly influential for the graph trend lines.
+It was then considered whether HTTPS and Traefik routing were the thing slowing down the simulator, but removing those things only made very slight improvements, so they were re-enabled.
+The actual reason turned out to be the server location, as will be described in [TODO: 4.2 Operations](42_Lessons-Operation.md).
+
+Without the graph as a motivator, these things would never have been discovered.
+
+#### Proper communication
+
+A problem that the team never got to improve on was communication, especially of the asynchronous kind.
+When co-located, the team managed to get lots of work done and reviewed, but at the times between co-located sessions and after the pandemic hit the world, then reviews were slow and impractical.
+
+A possible reason could be that reminding people to review is easier and more direct when working next to each other.
+Another could be that the other courses and work got in the way.
+
+A third possibility could be that there were no channels for politely nudging towards a review.
+The team tried to alleviate this with the `#review` channel on Discord, where anyone could push for a review.
+This helped in some ways, but simply pushed the notification into its own little box that was easier to forget or ignore.
```

Comment:
I disagree. The review channel meant I got way better at looking at reviews, because I was more likely to check discord than github, and Discord is clearer about notification source.

Reply:
Please see whether you agree with the update.